src/main/scala/is/hail/methods/MaximalIndependentSet.scala

```diff
@@ -40,42 +36,45 @@ object MaximalIndependentSet {
       Graph(toBeComputed.vertices.leftZipJoin(toBeComputed.degrees) { (v, _, degree) => (degree.getOrElse(0), v) }, toBeComputed.edges)
     }
 
-    var graph = updateVertexDegrees(g)
     val edgeDirection = if (undirected) EdgeDirection.Either else EdgeDirection.Out
 
-    while(graph.numEdges > 0) {
-      graph = graph.pregel(initialMsg, Int.MaxValue, edgeDirection)(receiveMessage, sendMsg, mergeMsg)
-      graph = graph.subgraph(_ => true, (id, value) => value match { case (maxDegrees, maxID) => maxID != id || maxDegrees == 0})
-      graph = updateVertexDegrees(graph)
-    }
-    graph.vertices.keys.collect().toSet
+    var idSetToDelete = mutable.Set[VertexId]()
+
+    g.persist()
+    var workingGraph = g
+
+    do {
+      idSetToDelete ++= updateVertexDegrees(workingGraph)
+          .pregel(initialMsg, Int.MaxValue, edgeDirection)(receiveMessage, sendMsg, mergeMsg)
+          .vertices
+          .filter{ case (id, (maxDegrees, maxID)) => maxID == id && maxDegrees != 0}
+          .map(_._1)
+          .collect()
+          .toSet
+      workingGraph.unpersist()
+      workingGraph = g.subgraph(_ => true, (id, value) => !idSetToDelete.contains(id))
+      workingGraph.persist()
+    } while (workingGraph.numEdges > 0)
```

Comment:
I think I lead you astray, I'm sorry! We actually want the condition at the top of the loop, i.e. a standard while loop. While this is correct, I think we end up with one unnecessary iteration if the graph is completely disconnected, right?